pkg/apis/storage/validation/validation.go

```diff
@@ -266,3 +279,176 @@ func validateAllowedTopologies(topologies []api.TopologySelectorTerm, fldPath *f
 
 	return allErrs
 }
+
+// ValidateCSINodeInfo validates a CSINodeInfo.
+func ValidateCSINodeInfo(csiNodeInfo *storage.CSINodeInfo) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&csiNodeInfo.ObjectMeta, false, apivalidation.ValidateNodeName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateCSINodeInfoSpec(&csiNodeInfo.Spec, field.NewPath("spec"))...)
+	allErrs = append(allErrs, validateCSINodeInfoStatus(&csiNodeInfo.Status, field.NewPath("status"))...)
+	return allErrs
+}
+
+// ValidateCSINodeInfoUpdate validates a CSINodeInfo.
+func ValidateCSINodeInfoUpdate(new, old *storage.CSINodeInfo) field.ErrorList {
+	allErrs := ValidateCSINodeInfo(new)
+
+	// Spec is read-only
+	// If this ever relaxes in the future, make sure to increment the Generation number in PrepareForUpdate
+	if !apiequality.Semantic.DeepEqual(old.Spec, new.Spec) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec"), new.Spec, "field is immutable"))
+	}
+	return allErrs
+}
+
+// ValidateCSINodeInfoSpec tests that the specified CSINodeInfoSpec
+// has valid data.
+func validateCSINodeInfoSpec(
+	spec *storage.CSINodeInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCSIDriverInfoSpecs(spec.Drivers, fldPath.Child("drivers"))...)
+	return allErrs
+}
+
+// ValidateCSINodeInfoStatus tests that the specified CSINodeInfoStatus
+// has valid data.
+func validateCSINodeInfoStatus(
+	status *storage.CSINodeInfoStatus, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, validateCSIDriverInfoStatuses(status.Drivers, fldPath.Child("drivers"))...)
+	return allErrs
+}
+
+// ValidateCSIDriverInfoSpecs tests that the specified CSIDriverInfoSpecs
+// have valid data.
+func validateCSIDriverInfoSpecs(
+	drivers []storage.CSIDriverInfoSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if drivers == nil || len(drivers) == 0 {
```

Comment:
len() == 0 captures nil -- no need to double test that.  same elsewhere

Reply:
Ok